### PR TITLE
refactor: convert node, circuitElement and modules to TypeScript

### LINKED ISSUE #1148: refactor: node.js, module.js, and circuitElement.js to TypeScript with type definitions
### Is there an existing issue for this?
- [x] I have searched the existing issues

### Description
Convert `node.js`, `module.js`, and `circuitElement.js` to `.ts`.

### Guidelines
- Rename `.js` → `.ts`, then run `vue-tsc --noEmit` and fix all resulting errors.
- No implicit `any` — add real types for params, returns, and class members.
- If `any`/`unknown` truly can't be removed, leave a one-line comment above it explaining why.
- Convert all three files in the same PR.
- Don't include file extensions in imports (`./node`, not `./node.ts`).
- Double check other files importing these three still typecheck after the change.

### PR DESCRIPTION
**Pull request by Mr-Daker**
Fixes #1148

#### Describe the changes you have made in this PR -

Converts `node.js`, `circuitElement.js` and `modules.js` to TypeScript in **both** `src/simulator/src/` and `v1/src/simulator/src/`. (The issue says `module.js`; the real file is `modules.js`.) The v1 copies were byte-identical to the v0 originals, so they take the same conversion and the same supporting changes.

**Key review point:** `useDefineForClassFields` is enabled, so a class field declaration — even `foo!: T` — emits an own property set to `undefined` that shadows the prototype. ~50 element modules set `objectType`, `alwaysResolve`, `rectangleObject`, `canShowInSubcircuit` etc. on their prototypes, so declaring those as fields would silently turn them all into `undefined`. Every prototype-backed member uses `declare`, which emits nothing. Verified by compiling the files and reading the emitted class bodies.

**Why other files changed** — each is forced by the new types, not cleanup. Both trees have their own copies, so each change is applied twice:

| File | Why |
|---|---|
| `eventQueue.ts` | `delay` was required, but the body is `delay \|\| obj.propagationDelay` and callers omit it → `delay?: number`. Also clears 2 pre-existing errors in `OrGate.ts`. |
| `interface/simulationArea.ts` | `hover` was typed `boolean` but holds the hovered element; `mouseXf` / `mouseYf` / `contentionPending` weren't declared at all. |
| `simulationArea.ts` | Initialises `contentionPending`, which `engine.js` replaces each simulation pass. |
| `circuit.ts` | One commented cast — the root element is built before `initialize()` creates the node/wire arrays. |

**Two behaviour notes:**
- `Node.prototype.processVerilog` was overwriting the class method of the same name with a different body (no `isVerilogResolvable()` guard). I kept the version that actually ran, so behaviour is unchanged — happy to switch if the guard was the intent.
- `queueProperties.time` / `index` start at `0` instead of `undefined`. Nothing outside `eventQueue.ts` reads them, and `saveObject()` output is byte-identical.

**Tests:** `prototypeIntegrity.spec.js` (6 tests, added to both trees) pins the prototype contract above. Verified red/green in each: swapping one `declare` back to `!` leaves all pre-existing tests passing — 177 in v0, 67 in v1 — while this spec fails with `expected undefined to be false`. Nothing else in either suite covers that class of regression.

**Verification** against `upstream/main` (`a5aa514`):

| Check | Base | This PR |
|---|---|---|
| `tsc --noEmit` (v0) | 390 errors | **379** |
| `tsc` on v1 (throwaway config) | 392 errors | **381** |
| errors in the converted files | n/a (were `.js`) | **0** in both trees |
| errors introduced in any file | — | **0** (no file's count increased) |
| `npm run lint` | exit 0, 23 warnings | exit 0, 23 warnings |
| `npm run lint:v1` (`--deny-warnings`) | exit 0 | exit 0 |
| `oxfmt --check` | clean | clean |
| `npm test` | 177 passed / 22 files | **189 passed / 24 files** (12 added) |
| `npm run build` (v0 + v1) | — | exit 0 |

Neither `tsc` nor any type-check runs in CI, and no tsconfig covers `v1/` at all — so those counts are local, measured against a baseline captured on `upstream/main`. I used a throwaway tsconfig for v1 rather than ship `.ts` nothing had compiled; it isn't part of this PR.

I also ran a throwaway differential harness — 29 probes over the paths this touches, captured on the original JS and on this branch. 26 of 29 identical; the 3 differences are all the `queueProperties` initial value above.

### Screenshots of the UI changes (If any) -

None — no UI changes.

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

`node.js` and `circuitElement.js` are the base classes every element module extends, so typing them is what unblocks converting the ~50 element files later. `modules.js` came with them because the element registry and `changeInputSize` are coupled to both. The v1 copies were byte-identical, so the same conversion applied — though nothing type-checks `v1/`, so I ran a throwaway tsconfig over it rather than trust that it compiled.

The hard part was `useDefineForClassFields`. I assumed `objectType!: string` would avoid the field emit; I compiled a test case and it doesn't — `!` only silences the initialisation check, the field is still emitted and still shadows the prototype. Only `declare` emits nothing, so every prototype-backed member uses it, while constructor-assigned fields use `!` (Node's constructor returns early on its `isNaN` guard).

Three things I rejected. **`||` for the `update` accumulator** — TypeScript suggests it, but it short-circuits, so once `update` is true the remaining `nodeList[i].update()` calls never happen; I kept a numeric accumulator, which is what the original already was since `|=` coerces to a number on first use. **Defaulting `mouseXf` to `0`** — tidier than optional fields plus assertions, but `scope.root` sits at (0,0) and would report itself hovered before the first pointer event, so the fields stay optional and the runtime values are untouched. **Importing the real `Scope` class** — its `allNodes` / `nodes` / `wires` are optional, needing ~20 non-null assertions in `node.ts`; a local interface plus one commented cast was smaller.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

## Summary by CodeRabbit

* **New Features**
  * Improved circuit-element interaction, including dragging, selection, layout editing, labels, sizing, and rendering.
  * Enhanced node wiring, simulation propagation, contention handling, and Verilog generation.
  * Input resizing now supports validated values while preserving selection.
  * Event scheduling supports custom or default propagation delays.

* **Bug Fixes**
  * Improved pointer-position handling and unconnected-output Verilog generation.

* **Tests**
  * Added coverage for element defaults, inheritance, resizing behavior, and Verilog output.

### CODE REVIEW

File: src/simulator/src/eventQueue.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ export class EventQueue {
     this.time = 0;
   }
 
-  add(obj: QueueObject, delay: number) {
+  add(obj: QueueObject, delay?: number) {
```
[the remark this answers is not in the document]
Comment (Mr-Daker) [reply src/simulator/src/eventQueue.ts]: Done — added `: void` in both trees.

File: src/simulator/src/interface/simulationArea.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import { EventQueue } from "../eventQueue";
+import ContentionPendingData from "../contention";
 export interface SimulationArea {
   canvas: HTMLCanvasElement;
   context: CanvasRenderingContext2D | null;
   selected: boolean;
-  hover: boolean;
+  // Holds the element or node currently under the pointer, or undefined.
+  // TODO: make this CircuitElement|Node|undefined once converted to typescript
+  hover: any;
```
[the remark this answers is not in the document]
Comment (Mr-Daker) [reply src/simulator/src/interface/simulationArea.ts]: Done, but with a wider union than suggested. `hover` doesn't only hold an element or a node — `engine.js:554` parks a `Wire` there, and `layout/layoutNode.ts:69` assigns a `LayoutNode`. Typing it as `CircuitElement | Node | undefined` would have reintroduced errors in `layoutNode.ts` that this PR fixes, so it's now:

`hover: CircuitElement | Node | Wire | LayoutNode | undefined`

One consequence worth flagging: the object literal started at `hover: false`, which isn't a member of that union, so it now starts `undefined`. `engine.js:302` gates the canvas cursor on `hover !== undefined`, so the cursor is `default` before the first hover instead of `pointer`. Happy to keep `| false` in the union instead if you'd rather this PR change no behaviour at all.

Also narrowed the one `simulationArea.hover.showHover` write with `instanceof Node` rather than a cast — only `Node` reads that flag, so it's equivalent to the previous unconditional write.

File: src/simulator/src/modules.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ * Global
+ */
+var modules: Record<string, ElementConstructor> = {};
+
+export default modules;
+
+export function changeInputSize(this: ResizableElement, size: number | string): unknown {
+  if (size == undefined || Number(size) < 2 || Number(size) > 10) return;
+  if (this.inputSize == Number(size)) return;
+  const newSize = parseInt(String(size), 10);
```
[the remark this answers is not in the document]
Comment (Mr-Daker) [reply src/simulator/src/modules.ts]: Done. `Number.isInteger` check up front, so `"abc"` and `2.5` are rejected instead of reaching `parseInt`. Note this is a behaviour change rather than a typing fix — the original genuinely built elements with `NaN` inputs — so I added tests and verified them red/green: reverting the guard makes the new test fail with `expected OrGate{...} to be undefined`.

**Comment by Harkeerat24** [comment]
@coderabbitai Review this PR for idiomatic typescript, type safety, potential bugs when types are not the same, potential runtime errors etc

**Comment by Mr-Daker** [comment]
Thanks — I checked both, and pushed a fix for the comment that caused the first one.

**getNextPosition / Output overlap:** I don't think this one holds. The two callers pass different `x` — `Input.js:38` uses `getNextPosition(0, scope)`, `Output.js:38` uses `getNextPosition(scope.layout.width, scope)` — and both loops only mark a Y when `layoutProperties.x === x`. Inputs live at `x = 0`, Outputs at `x = scope.layout.width`, so when positioning an Output the x check already excludes every Input, skipped or not. Verified locally:

    LAYOUT_WIDTH 100
    INPUTS   [[0,20],[0,40],[0,60]]
    OUTPUTS  [[100,20],[100,40]]
    SAME_XY_COLLISION false

The Y values do coincide, but on opposite sides of the subcircuit box — that's the intended layout.

You were right that the comment was misleading though: it only explained Input's constructor, which made the skip look unsafe for Output's. Rewritten in f915b82 to cover both cases.

**Type-check coverage:** `v0/src` is a symlink to `../src`, so `"v0/src/**/*"` does cover `src/simulator/**`. `tsc --listFiles` lists `v0/src/simulator/src/{node,modules,circuitElement}.ts`, and 369 of the 379 reported errors sit under `v0/src/simulator/` — that's how the numbers in the description were measured.

You're right about the other half: nothing covers `v1/`, and there's no `tsc` script in CI. Both are noted in the description, and it's why I ran a throwaway config over v1. Adding a CI type-check would be worth doing, but it can't land in this PR — it would fail immediately on the 379 + 382 errors that predate it. Happy to open a separate issue for that.

---

@coderabbitai full review — please re-check for idiomatic TypeScript, type safety, mismatched types across the v0/v1 trees, and potential runtime errors.

**Comment by Mr-Daker** [comment]
@Harkeerat24 Can you please review it?

**Review by aryanndwi123 — APPROVED** [review]
LGTM 🚀
